opal/mca/btl/iwarp/btl_iwarp.c

```diff
@@ -0,0 +1,2009 @@
+        if(NULL != values.receive_queues) {
+            if(0 != strcmp(mca_btl_iwarp_component.receive_queues,
+                           values.receive_queues)) {
+                opal_show_help("help-mpi-btl-iwarp.txt",
+                               "unsupported queues configuration", true,
+                               opal_process_info.nodename,
+                               ibv_get_device_name(iwarp_btl->device->ib_dev),
+                               (iwarp_btl->device->ib_dev_attr).vendor_id,
+                               (iwarp_btl->device->ib_dev_attr).vendor_part_id,
+                               mca_btl_iwarp_component.receive_queues,
+                               opal_get_proc_hostname(endpoint->endpoint_proc->proc_opal),
+                               endpoint->rem_info.rem_vendor_id,
+                               endpoint->rem_info.rem_vendor_part_id,
+                               values.receive_queues);
+
+                return OPAL_ERROR;
+            }
+        }
+        break;
+    }
+
+    return OPAL_SUCCESS;
+}
+
+static int prepare_device_for_use (mca_btl_iwarp_device_t *device)
+{
+    mca_btl_iwarp_frag_init_data_t *init_data;
+    int rc = OPAL_SUCCESS, length;
+
+    opal_mutex_lock(&device->device_lock);
+
+    if (device->ready_for_use) {
+        goto exit;
+    }
+
+    /* For each btl module that we made - find every
+       base device that doesn't have device->qps setup on it yet (remember
+       that some modules may share the same device, so when going through
+       to loop, we may hit a device that was already setup earlier in
+       the loop).
+
+       We may to call for prepare_device_for_use() only after adding the btl
+       to mca_btl_iwarp_component.iwarp_btls, since the prepare_device_for_use
+       adds device to async thread that require access to
+       mca_btl_iwarp_component.iwarp_btls.
+    */
+
+    /* Setup the device qps info */
+    device->qps = (mca_btl_iwarp_device_qp_t*)
+        calloc(mca_btl_iwarp_component.num_qps,
+               sizeof(mca_btl_iwarp_device_qp_t));
+    if (NULL == device->qps) {
+        BTL_ERROR(("Failed malloc: %s:%d", __FILE__, __LINE__));
+        rc = OPAL_ERR_OUT_OF_RESOURCE;
+        goto exit;
+    }
+
+    for (int qp_index = 0 ; qp_index < mca_btl_iwarp_component.num_qps ; qp_index++) {
+        OBJ_CONSTRUCT(&device->qps[qp_index].send_free, opal_free_list_t);
+        OBJ_CONSTRUCT(&device->qps[qp_index].recv_free, opal_free_list_t);
+    }
+
+    device->got_fatal_event = false;
+    device->got_port_event = false;
+    mca_btl_iwarp_async_add_device (device);
+
+#if OPAL_ENABLE_PROGRESS_THREADS == 1
+    /* Prepare data for thread, but not starting it */
+    OBJ_CONSTRUCT(&device->thread, opal_thread_t);
+    device->thread.t_run = mca_btl_iwarp_progress_thread;
+    device->thread.t_arg = device;
+    device->progress = false;
+#endif
+
+#if HAVE_XRC
+    /* if user configured to run with XRC qp and the device doesn't
```

Comment:
Delete all XRC.  I'll stop commenting on. it -- I see XRC blocks in several other places in this file.